File: src/Pawchums.Domain/Entities/RescueInitiation/RescueInitiation.cs
using Pawchums.Entities.RequestRescues;
using System;
using Volo.Abp.Domain.Entities.Auditing;
using Volo.Abp.MultiTenancy;

namespace AnimalRescueSystem.Entities.RequestRescues;

/// <summary>
/// Tracks when rescuers initiate/accept a rescue request
/// Multiple rescuers can express interest, but only one is selected
/// </summary>
public class RescueInitiation : FullAuditedEntity<Guid>, IMultiTenant
{
    public Guid? TenantId { get; set; }

    // Foreign Keys
    public virtual Guid RequestRescueId { get; set; }
    public virtual Guid RescuerId { get; set; } // Foreign key to AbpUsers.Id

    // Initiation Details
    public virtual DateTime InitiatedDate { get; set; }
    public virtual string? Notes { get; set; } // Rescuer's notes when initiating
    public virtual string Status { get; set; } // Pending, Accepted, Rejected, Withdrawn
    public virtual bool IsSelected { get; set; } // True if this rescuer was selected for the job
    public virtual DateTime? AcceptedDate { get; set; } // When admin accepted this rescuer
    public virtual Guid? AcceptedByUserId { get; set; } // Admin who accepted

    // Navigation Properties
    public virtual RequestRescue RequestRescue { get; set; }
}

File: src/types/rescueInitiation.ts
// Rescue Initiation DTOs and Types

export interface CreateRescueInitiationDto {
  requestRescueId: string;
  notes?: string;
}

export interface UpdateRescueInitiationDto {
  notes?: string;
}

export interface RescueInitiationDto {
  id: string;
  requestRescueId: string;
  requestTitle: string;
  requestLocation: string;
  rescuerId: string;
  rescuerName?: string;
  rescuerEmail?: string;
  initiatedDate: string;
  notes?: string;
  status: RescueInitiationStatus;
  isSelected: boolean;
  acceptedDate?: string;
  acceptedByUserId?: string;
  acceptedByUserName?: string;
  creationTime: string;
}

export interface RescueInitiationResponseDto {
  id: string;
  requestRescueId: string;
  status: string;
}

export interface RescueInitiationFilter {
  searchKeyword?: string;
  status?: RescueInitiationStatus;
  isSelected?: boolean;
  requestRescueId?: string;
  rescuerId?: string;
}

export type RescueInitiationStatus =
  | 'Pending'
  | 'Accepted'
  | 'Rejected'
  | 'Withdrawn';

export const RESCUE_INITIATION_STATUS_OPTIONS = [
  { label: 'Pending', value: 'Pending' },
  { label: 'Accepted', value: 'Accepted' },
  { label: 'Rejected', value: 'Rejected' },
  { label: 'Withdrawn', value: 'Withdrawn' },
] as const;
